odl/operator/pspace_ops.py

```diff
@@ -562,6 +628,286 @@ def adjoint(self):
+            Input vector that should be used in the reduction
+        out : range element, optional
+            output vector to write result to
+
+
+        Parameters
+        ----------
+        x : domain element
+            Input vector to be evaluated by operators
+        out : range element, optional
+            output vector to write result to
+
+        Returns
+        -------
+        out : range element
+            Sum of operators evaluated in point
+
+        Examples
+        --------
+        >>> import odl
+        >>> I = odl.IdentityOperator(odl.Rn(3))
+        >>> op = ReductionOperator(I, 2 * I)
+        >>> op([[1.0, 2.0, 3.0], [4.0, 6.0, 8.0]])
+        Rn(3).element([9.0, 14.0, 19.0])
+        """
+        if out is None:
+            return self.prod_op(x)[0]
+        else:
+            wrapped_out = self.prod_op.range.element([out], cast=False)
+            return self.prod_op(x, out=wrapped_out)
+
+    def derivative(self, x):
+        """Derivative of the reduction operator.
+
+        Parameters
+        ----------
+        x : domain element
+            The point to take the derivative in
+
+        Returns
+        -------
+        adjoint : linear `BroadcastOperator`
+            The derivative
+
+        Examples
+        --------
+        >>> import odl
+        >>> r3 = odl.Rn(3)
+        >>> I = odl.IdentityOperator(r3)
+        >>> x = r3.element([1.0, 2.0, 3.0])
+        >>> y = r3.element([4.0, 6.0, 8.0])
+
+        Example with linear operator (derivative is itself)
+
+        >>> op = ReductionOperator(I, 2 * I)
+        >>> op([x, y])
+        Rn(3).element([9.0, 14.0, 19.0])
+        >>> op.derivative([x, y])([x, y])
+        Rn(3).element([9.0, 14.0, 19.0])
+
+        Example with affine operator
+
+        >>> residual_op = odl.ResidualOperator(I, r3.element([1, 1, 1]))
+        >>> op = ReductionOperator(residual_op, 2 * residual_op)
+
+        Calling operator gives offset by [3, 3, 3]
+
+        >>> op([x, y])
+        Rn(3).element([6.0, 11.0, 16.0])
+
+        Derivative of affine operator does not have this offset
+
+        >>> op.derivative([x, y])([x, y])
+        Rn(3).element([9.0, 14.0, 19.0])
+        """
+        return ReductionOperator(*[op.derivative(xi)
+                                   for op, xi in zip(self.operators, x)])
+
+    @property
+    def adjoint(self):
+        """Adjoint of the reduction operator.
+
+        Returns
+        -------
+        adjoint : linear `BroadcastOperator`
+            The adjoint
+
+        Examples
+        --------
+        >>> import odl
+        >>> I = odl.IdentityOperator(odl.Rn(3))
+        >>> op = ReductionOperator(I, 2 * I)
+        >>> op.adjoint([1, 2, 3])
+        ProductSpace(Rn(3), 2).element([
+            [1.0, 2.0, 3.0],
+            [2.0, 4.0, 6.0]
+        ])
+        """
+        return BroadcastOperator(*[op.adjoint for op in self.operators])
+
+
+def diagonal_operator(operators, dom=None, ran=None):
```

Comment:
I see. Then I think ordering them beforehand would make sense IMO. If 
supporting different orderings is that simple, then we should do it like 
that. Perhaps with a dedicated method that reorders if a different 
ordering is desired? Internally it can always be 'C'.

On 02/09/2016 01:48 PM, Jonas Adler wrote:
> In odl/operator/pspace_ops.py
> <https://github.com/odlgroup/odl/pull/178#discussion_r52302675>:
>
>> +
>> +        Examples
>> +        --------
>> +        >>> import odl
>> +        >>> I = odl.IdentityOperator(odl.Rn(3))
>> +        >>> op = ReductionOperator(I, 2 * I)
>> +        >>> op.adjoint([1, 2, 3])
>> +        ProductSpace(Rn(3), 2).element([
>> +            [1.0, 2.0, 3.0],
>> +            [2.0, 4.0, 6.0]
>> +        ])
>> +        """
>> +        return BroadcastOperator(*[op.adjoint for op in self.operators])
>> +
>> +
>> +def diagonal_operator(operators, dom=None, ran=None):
>
>     I'm not so familiar with the internals of ProductSpaceOperator, but
>     I thought they were stored in a matrix. Then there should be a
>     well-defined way to get them back, using the matrix ordering maybe?
>     I don't think supporting different orderings is necessary. If we
>     write that we always store in 'C' order that should be fine?
>
> It is stored in a coo_matrix
> <http://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.sparse.coo_matrix.html>,
> to the best of my knowledge, this means that each element is stored as a
> tripple (col, row, value), thus it can have basically any ordering. All
> methods that we currently use to access thing work on this tripple.
>
> Example storing in "reverse" order
>
>>> >  A  =  scipy.sparse.coo_matrix(([2,1], ([1,0], [1,0])))
>>> >  A.data
> array([2,1])
>>> >  A.todense()
> matrix([[1,0],
>          [0,2]])
>
>     Wasn't there a property to get the matrix? Maybe there should be
>     something like asmatrix() then.
>
> There is no such property, but it is stored as a nonprivate member
> |ops|, we should probably expose it.
>
> One option would be to somehow sort the array:
>
>>> >  result=  sorted(zip(col, row, data))# swap col and row to change order
>>> >  result
> [(0,0,1), (1,1,2)]
>>> >  col, row, values=  zip(*result)# sorted
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/odlgroup/odl/pull/178/files#r52302675>.
>